qiskit/transpiler/passes/mapping/algorithm/dependency_graph.py

```diff
@@ -0,0 +1,323 @@
+                    elif gate.name in z_gates or (gate.name == "cx" and gate.qargs[0] == tbit):
+                        self._graph.add_edge(m, n)
+                        z_flag = True
+                    else:
+                        raise TranspilerError("Unknown gate: " + gate.name)
+
+                x_flag = False
+                for m in cpgow:  # control bit: bc
+                    gate = self._gates[m]
+                    if gate.name in b_gates:
+                        self._graph.add_edge(m, n)
+                        break
+                    elif gate.name in x_gates or (gate.name == "cx" and gate.qargs[1] == cbit):
+                        self._graph.add_edge(m, n)
+                        x_flag = True
+                    elif gate.name in z_gates or (gate.name == "cx" and gate.qargs[0] == cbit):
+                        if x_flag:
+                            break
+                        else:
+                            continue
+                    else:
+                        raise TranspilerError("Unknown gate: " + gate.name)
+            elif self._gates[n].name in b_gates:
+                for i, pgow in enumerate(self._prior_gates_on_wire(self._gates, n)):
+                    b = self._all_args(n)[i]
+                    x_flag, z_flag = False, False
+                    for m in pgow:
+                        gate = self._gates[m]
+                        if gate.name in b_gates:
+                            self._graph.add_edge(m, n)
+                            break
+                        elif gate.name in x_gates or (gate.name == "cx" and gate.qargs[1] == b):
+                            if z_flag:
+                                break
+                            else:
+                                self._graph.add_edge(m, n)
+                                x_flag = True
+                        elif gate.name in z_gates or (gate.name == "cx" and gate.qargs[0] == b):
+                            if x_flag:
+                                break
+                            else:
+                                self._graph.add_edge(m, n)
+                                z_flag = True
+                        else:
+                            raise TranspilerError("Unknown gate: " + gate.name)
+            else:
+                raise TranspilerError("Unknown gate: " + self._gates[n].name)
+
+    def _create_basic_graph(self):
+        for n in self._graph.nodes():
+            for pgow in self._prior_gates_on_wire(self._gates, n):
+                m = next(pgow, -1)
+                if m != -1:
+                    self._graph.add_edge(m, n)
+
+    def n_nodes(self) -> int:
+        """Number of the nodes
+        Returns:
+            Number of the nodes in this graph.
+        """
+        return self._graph.__len__()
+
+    def qargs(self, i: int) -> List[Qubit]:
+        """Qubit arguments of the gate
+        Args:
+            i: Index of the gate in the `self._gates`
```

Comment:
In this API, DependencyGraph depends on QuantumCircuit (say `qc`) and all of the gates are considered as integers, which are the indices of `qc.data`. Since DependencyGraph is constructed from a QuantumCircuit, user knows all `_gates` as `qc.data`. All public functions (except for `gates`) accept and return integers. I might set gate objects directly as nodes of graph instead of integers, but I didn't do so, just because object id can be changed even though it would be a very very rare case.
(Excuse me for my confusing variable names `gate`/`gates`, in this case I should use instruction instead of gate, or more precisely `instruction_context`, because they include non-gate instructions like measure and their acting (qu)bits. I had followed naming convention in old-version qiskit, but what `gate` means was changed at some point. Should I refactor all names around `gate`?)

Reply:
Thank you! Could you put just a little of the essence of that description into the docstrings, so that other people with the same question might understand when they come across this? 

If you think there is a better name, I am all for updating it! :D Don't be afraid to use a better name when you think of it! And it's easier to change the sooner you do so :)

Reply:
Update docstrings at 4fd0103